File: packages/runtime/src/camera.ts
/**
 * The speaker's camera, filling the tile the deck declared.
 *
 * The feature is two opt-ins and refuses to work on one of them.
 *
 * **The author's**, in the file: `camera: side` puts an empty tile in a named
 * region of the slide. That is all the build emits — no video element, no
 * script, nothing that could reach a device. A published deck is a static page
 * somebody opens from a link, a QR code, or an archive years later, and a page
 * that asks for a webcam is a page people close.
 *
 * **The speaker's**, at presentation time: this module is reached from
 * `enterPresentation`, which happens on an explicit gesture and nowhere else.
 * A page nobody is presenting from never calls it, so a reader is never asked.
 *
 * `startCamera` on a document with no tile in it returns without touching
 * `getUserMedia` at all. The gate is structural rather than remembered: there
 * is no order of calls that opens a camera on a deck whose author did not place
 * one.
 *
 * # Failure is ordinary
 *
 * No camera, permission refused, and a camera another application is already
 * holding all happen minutes before a talk. None of them is exceptional and
 * none of them may cost the deck, so nothing here rejects: the tile ends up
 * saying what went wrong, in the speaker's words, on the slide they are looking
 * at. A tile reading "no camera found" is better than a black rectangle, and
 * far better than an exception thrown into a slide.
 *
 * # Nothing leaves the machine
 *
 * The stream reaches exactly one place — the `<video>` element this module puts
 * in the tile — and there is no way to get it back out. That is a camera on the
 * speaker's own screen, which is what the audience is already looking at, and
 * not a broadcast.
 */

/** Attribute naming the region a tile occupies. Mirrors `slidx_core::camera`. */
export const CAMERA_ATTRIBUTE = "data-slidx-camera";

/** Attribute carrying what the camera is doing. Mirrors `slidx_core::camera`. */
export const CAMERA_STATE_ATTRIBUTE = "data-slidx-camera-state";

/**
 * What the camera actually is, which is rarely what was asked for.
 *
 * Every value except `live` leaves the deck presentable. They are separate
 * rather than one `failed` because the speaker's next action differs: a refused
 * permission is a browser setting, a busy device is an application to quit, and
 * an absent one is nothing they can do anything about in the two minutes they
 * have.
 */
export type CameraStatus =
  /** The deck asked for none, or nobody started one. */
  | "off"
  | "live"
  /** The speaker, or the browser's stored decision, said no. */
  | "denied"
  /** No camera on this machine. */
  | "absent"
  /** Something else has it — usually the conferencing app for the same talk. */
  | "busy"
  /** This browser has no camera API to ask. */
  | "unsupported";

/** What presentation mode got, and how to give it back. */
export interface CameraSession {
  readonly status: CameraStatus;
  /** One sentence, written for a speaker rather than for a log. */
  readonly detail: string;
  /** Releases the device. Safe to call twice. */
  stop(): void;
}

/** Only the part of a stream this module can use, so a test can supply one. */
export interface MediaStreamLike {
  getTracks(): { stop(): void }[];
}

/**
 * Injected so this is testable, and so a missing API is a value not a crash.
 *
 * One capability, deliberately. There is nowhere for a stream to go except the
 * element on the slide.
 */
export interface CameraEnvironment {
  /** Absent when the browser has no camera API at all. */
  requestStream?: () => Promise<MediaStreamLike>;
}

/** The real browser, for callers that are not a test. */
export function browserCameraEnvironment(view: Window = globalThis.window): CameraEnvironment {
  const devices = (view.navigator as Navigator & { mediaDevices?: MediaDevices }).mediaDevices;

  // Spread rather than assigned `undefined`, so a browser without the API
  // genuinely leaves the key off and `startCamera` reports `unsupported`
  // instead of failing somewhere less legible.
  return {
    ...(devices === undefined
      ? {}
      : {
          // Video only. A talk's audio comes from the room's microphone, and
          // asking for a device this feature does not use is asking a speaker
          // to grant a permission for nothing.
          requestStream: () => devices.getUserMedia({ video: true, audio: false }),
        }),
  };
}

/** A session that holds nothing, for every path that never opened a device. */
function idle(status: CameraStatus, detail: string): CameraSession {
  return { status, detail, stop: () => {} };
}

/**
 * Fills the tile this document declares, if it declares one.
 *
 * Never rejects. Called from `enterPresentation`, alongside the wake lock and
 * fullscreen, and reports what it got the same way they do.
 */
export async function startCamera(
  root: Document,
  environment: CameraEnvironment,
): Promise<CameraSession> {
  const tile = root.querySelector(`[${CAMERA_ATTRIBUTE}]`);

  // The author's opt-in, enforced before anything else can happen. A deck with
  // no camera in it has no path from here to a device.
  if (tile === null) return idle("off", "this slide places no camera");

  if (!environment.requestStream) {
    return show(tile, "unsupported", "this browser has no camera API");
  }

  let stream: MediaStreamLike;
  try {
    stream = await environment.requestStream();
  } catch (error) {
    const [status, detail] = describe(error);
    return show(tile, status, detail);
  }

  const video = root.createElement("video");
  video.autoplay = true;
  // Muted and inline are not preferences. An unmuted self-view feeds the room's
  // own speakers back into the room, and a browser will refuse to autoplay it
  // anyway — so the tile would sit on a play button in front of an audience.
  video.muted = true;
  video.playsInline = true;
  // The DOM types `srcObject` as the platform's whole `MediaStream`. Taking
  // that interface here would make the module untestable without a device, and
  // the only thing it ever does with a stream is stop its tracks.
  video.srcObject = stream as unknown as MediaProvider;

  tile.replaceChildren(video);
  tile.setAttribute(CAMERA_STATE_ATTRIBUTE, "live");

  let stopped = false;

  return {
    status: "live",
    detail: "the camera is on the slide",

    stop() {
      // Exiting twice is normal: a keyboard shortcut, Escape, and the browser's
      // own fullscreen control all land in `exit`. A track stopped twice is
      // harmless, but a tile emptied twice would clear a later session's video.
      if (stopped) return;
      stopped = true;

      for (const track of stream.getTracks()) track.stop();

      // Back to the state the build emitted, which draws as nothing. Left as
      // `live` the slide would keep a frozen last frame of the speaker on it.
      tile.replaceChildren();
      tile.setAttribute(CAMERA_STATE_ATTRIBUTE, "idle");
    },
  };
}

/**
 * Puts the reason on the slide and hands back a session that holds nothing.
 *
 * The speaker is looking at the slide, so that is where this belongs. `slidx
 * doctor` is the other half and the better one — it says the same thing before
 * anyone is on stage — but a permission can be refused after the pre-flight and
 * a camera can be taken by a conferencing app in the minute between them.
 */
function show(tile: Element, status: CameraStatus, detail: string): CameraSession {
  const message = tile.ownerDocument.createElement("p");
  message.className = "slidx-camera-status";
  message.textContent = detail;

  tile.replaceChildren(message);
  tile.setAttribute(CAMERA_STATE_ATTRIBUTE, status);

  return idle(status, detail);
}

/**
 * What a refusal actually was.
 *
 * The names are the ones `getUserMedia` rejects with, and each maps to a
 * different next action. Anything unrecognised is `unsupported` rather than
 * `denied`: telling a speaker to go and change a permission that was never the
 * problem costs them the minute they had.
 */
function describe(error: unknown): [CameraStatus, string] {
  const name = error instanceof Error ? error.name : "";

  switch (name) {
    case "NotAllowedError":
    case "SecurityError":
      return ["denied", "camera permission was refused for this page"];
    case "NotFoundError":
    case "OverconstrainedError":
      return ["absent", "no camera found on this machine"];
    case "NotReadableError":
    case "AbortError":
      return ["busy", "another application is using the camera"];
    default:
      return ["unsupported", "the camera could not be started"];
  }
}

File: crates/slidx_core/src/camera.rs
//! The speaker's own camera, as the deck asks for it.
//!
//! A talk given remotely, or recorded, wants the speaker's face on the slide.
//! The mechanism is a webcam, and a webcam is the one thing on this list that a
//! published deck must never reach for on its own: a deck is a static page
//! somebody opens from a link, a QR code, or an archive years later, and a page
//! that asks for a camera is a page people close.
//!
//! So the declaration and the stream are two different things, and only the
//! first of them is in the file. What an author writes here says *where* on the
//! slide a camera belongs. Nothing in this crate, and nothing the build emits
//! from it, opens a device — that needs a second opt-in, from the speaker, at
//! presentation time, and it lives in the runtime.
//!
//! # Why a region and not a corner
//!
//! `camera: side` names a region of the slide's layout, for the same reason a
//! block does: four floats would mean a different thing at a different aspect
//! ratio, no reviewer could read them, and no rule could measure them. A region
//! is a share of the slide at every projector size, and it is what lets a talk
//! put the speaker large on the opening slide and small over a diagram without
//! anyone typing a pixel.
//!
//! There is deliberately no `camera: true`. A camera with no region would have
//! to default to the region the slide's content is already in, which is the one
//! placement an author never means.

use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

use crate::diagnostic::{Diagnostic, Diagnostics};

/// Attribute naming the region a camera tile occupies.
///
/// The placement, and nothing about a device. It is written by the build into a
/// tile that starts empty, so a reader who opens the deck from a link gets an
/// element with no stream in it and no code that could ask for one.
pub const CAMERA_ATTRIBUTE: &str = "data-slidx-camera";

/// Attribute carrying what the camera is currently doing.
///
/// Written by the runtime and never by the build, which is what makes the
/// starting state — `idle`, drawn as nothing at all — the state every published
/// page is stuck in.
pub const CAMERA_STATE_ATTRIBUTE: &str = "data-slidx-camera-state";

/// A camera, as the deck declares it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Camera {
    /// The layout region the tile occupies, as the author wrote it.
    ///
    /// Kept as written rather than resolved: which regions exist belongs to a
    /// theme, and this crate has never heard of one. A name no layout has is
    /// reported by `slidx_theme::layout`, exactly as a misplaced block is.
    pub region: String,
}

/// Reads a `camera:` declaration.
///
/// Two levels of [`Option`], and both are load-bearing. The outer one says
/// whether the block mentioned `camera` at all, and the inner one says whether
/// it asked for one — without that distinction `camera: false` on a slide could
/// not switch off a deck-wide default, because a slide turning the camera off
/// would look exactly like a slide that never mentioned it.
pub fn parse(matter: &JsonValue, diagnostics: &mut Diagnostics) -> Option<Option<Camera>> {
    let field = crate::frontmatter::field(matter, "camera")?;

    // `camera: false` is how YAML reads the switch an author reaches for, and it
    // has to keep meaning "off" rather than falling through to the report below
    // — a slide that lost it would inherit the very camera it was turning off.
    if field == &JsonValue::Bool(false) {
        return Some(None);
    }

    match field.as_str().map(str::trim) {
        Some("none" | "off") => Some(None),
        Some(region) if !region.is_empty() => Some(Some(Camera { region: region.to_string() })),
        // Anything else — `camera: true`, a number, an empty string. Reported
        // rather than dropped: an author looking at a slide with no camera on it
        // needs to be told which line is the reason.
        _ => {
            diagnostics.push(
                Diagnostic::warning("frontmatter/invalid-camera", "`camera` must name a region")
                    .with_help("write `camera: side`, or `camera: false` to switch one off"),
            );

            Some(None)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn read(value: &JsonValue) -> (Option<Option<Camera>>, Diagnostics) {
        let mut diagnostics = Diagnostics::default();
        let camera = parse(value, &mut diagnostics);
        (camera, diagnostics)
    }

    #[test]
    fn a_camera_names_the_region_of_the_slide_it_occupies() {
        let (camera, diagnostics) = read(&json!({ "camera": "side" }));

        assert_eq!(camera, Some(Some(Camera { region: "side".into() })));
        assert!(diagnostics.is_empty());
    }

    #[test]
    fn a_slide_that_says_nothing_about_a_camera_is_left_free_to_inherit_one() {
        // The outer `None`. A slide with no opinion takes the deck's, which is
        // how a remote talk declares the camera once instead of forty times.
        assert_eq!(read(&json!({ "title": "Not a camera" })).0, None);
    }

    #[test]
    fn a_slide_can_switch_off_a_deck_wide_camera() {
        // The one slide that is a full-bleed diagram, in a talk that otherwise
        // keeps the speaker on screen.
        for value in [json!({ "camera": false }), json!({ "camera": "none" })] {
            let (camera, diagnostics) = read(&value);

            assert_eq!(camera, Some(None), "{value} did not switch the camera off");
            assert!(diagnostics.is_empty());
        }
    }

    #[test]
    fn a_camera_with_no_region_is_reported_rather_than_placed_over_the_slide() {
        // `camera: true` has only one region left to fall back to, and it is
        // the one the slide's own content is in.
        let (camera, diagnostics) = read(&json!({ "camera": true }));

        assert_eq!(camera, Some(None));
        assert_eq!(diagnostics.as_slice()[0].code, "frontmatter/invalid-camera");
        assert!(!diagnostics.has_blocking(), "the deck still renders");
    }

    #[test]
    fn a_camera_that_is_not_a_name_is_reported() {
        let (camera, diagnostics) = read(&json!({ "camera": 3 }));

        assert_eq!(camera, Some(None));
        assert_eq!(diagnostics.as_slice()[0].code, "frontmatter/invalid-camera");
    }

    #[test]
    fn the_region_name_is_taken_as_written_rather_than_resolved_here() {
        // Which regions exist belongs to a theme. A name this crate filtered out
        // would be a typo nothing downstream could report.
        let (camera, diagnostics) = read(&json!({ "camera": "  nowhere  " }));

        assert_eq!(camera, Some(Some(Camera { region: "nowhere".into() })));
        assert!(diagnostics.is_empty());
    }
}
